### DISCUSSION TITLE
Problem with vTextureCoord in shaders

### DISCUSSION THREAD

**Discussion by kvanonavk** [pixijs/pixijs#7773] [Q&A] [answered]
Hi, everyone! I'm trying to create a very simple shader which will visualize UVs on a rect (Graphics) and faced a strange problem with vTextureCoord. When I render a rect whose size is greater than 0.5 of application render area (Application.screen) vTextureCoord is calculated in a strange way. It's like vTextureCoord ignores rect and is calculated based solely on screen size. Here are the screenshots and source code.

JS code:

```
this.app = new PIXI.Application({ width: 512, height: 512 });

let drawRect = this.app.screen;
drawRect.width *= 0.5;
drawRect.height *= 0.5;
//drawRect.width *= 0.51;
//drawRect.height *= 0.51;

let background = new PIXI.Graphics();
background.drawRect(drawRect.x, drawRect.y, drawRect.width, drawRect.height);

let filter = new PIXI.Filter(undefined, fragSource);
background.filters = [filter];

this.app.stage.addChild(background);

document.body.appendChild(this.app.view);
```

Fragment shader code:
```
void main(void) {
	gl_FragColor.rg = vTextureCoord;
}
```

Screenshot when rect's size is 0.5 of screen's size:
![image](https://user-images.githubusercontent.com/11180220/132002939-06062851-93ee-49c8-8504-1ae2372ea91b.png)

Screenshot when rect's size is 0.51 of screen's size:
![image](https://user-images.githubusercontent.com/11180220/132002987-b8a03b69-3d52-47bb-9df9-6932722cd64f.png)

Is this some kind a of bug or am I missing something?

**Comment by ivanpopelyshev** [comment, marked as answer]
Temporary framebuffers that pixi uses for filters have pow2 size, not exact filterArea size

vTextureCoord is in normalized input coord, but you need normalized filter coord.

https://github.com/pixijs/pixijs/wiki/v5-Creating-filters#conversion-functions

Its really strange because I post this thing in every thread and people are still failing to notice it :)

**Reply by ivanpopelyshev** [reply]
Also, do you really need Filter? maybe its better to use mesh-shader

https://pixijs.io/examples/#/mesh-and-shaders/triangle-textured.js
https://pixijs.io/examples/#/mesh-and-shaders/shadertoy-mesh.js

**Reply by kvanonavk** [reply]
Hm, okay, thank you. I'll take a look.

**Reply by kvanonavk** [reply]
mesh-shader is exactly what I was looking for. Thanks again!
